<a href="https://colab.research.google.com/github/Ravitejatalluri/Hands-on-2/blob/main/Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

HANDS-ON SESSION-II: DATA MINING TOOLS

Name: RAVI TEJA TALLURI

REGRESSION MODELS

IMPORT THE LIBRARIES

In [15]:

import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import Ridge, Lasso, ElasticNet
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_squared_error
import numpy as np

IMPORT THE DATA USING PANDAS

In [2]:
data_realest = pd.read_csv('/content/realest.csv')
data_realest

Price  Bedroom   Space  Room   Lot     Tax  Bathroom  Garage  Condition
0     53.0      2.0   967.0   5.0  39.0   652.0       1.5     0.0        0.0
1     55.0      2.0   815.0   5.0  33.0  1000.0       1.0     2.0        1.0
2     56.0      3.0   900.0   5.0  35.0   897.0       1.5     1.0        0.0
3     58.0      3.0  1007.0   6.0  24.0   964.0       1.5     2.0        0.0
4     64.0      3.0  1100.0   7.0  50.0  1099.0       1.5     1.5        0.0
..     ...      ...     ...   ...   ...     ...       ...     ...        ...
152   44.0      3.0   820.0   5.0  27.0     NaN       1.0     0.0        1.0
153   43.0      2.0   593.0   4.0  30.0   465.0       2.0     1.0        0.0
154   53.0      3.0   716.0   6.0  30.0   585.0       1.0     2.0        1.0
155   62.0      4.0   951.0   7.0  30.0   895.0       2.0     1.0        0.0
156    NaN      NaN     NaN   NaN   NaN     NaN       NaN     NaN        NaN

[157 rows x 9 columns]

Checking Null Values

In [3]:
null_miss=data_realest.isnull().sum()
null_miss


Price         1
Bedroom       1
Space        11
Room          1
Lot          11
Tax          10
Bathroom      1
Garage        1
Condition     1
dtype: int64

Handling the missing values


In [6]:
# Drop rows where 'Price' is missing, as it is our target variable
data_realest = data_realest.dropna(subset=['Price'])

# Impute missing values in other columns with the median of each column
data_realest.fillna(data_realest.median(), inplace=True)

# Verify there are no remaining missing values
missing_values = data_realest.isnull().sum()
missing_values



Price        0
Bedroom      0
Space        0
Room         0
Lot          0
Tax          0
Bathroom     0
Garage       0
Condition    0
dtype: int64

The null value from price column was removed. the remaining columns null values were replaced with the median of the remaining data.

Split the data into training and testing for regression.

In [17]:
X = data_realest.drop(columns=['Price'])
y = data_realest['Price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.28, random_state=42)

Linear Regression Model

In [18]:
# Initialize and train the Linear Regression model
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_linear = linear_model.predict(X_test)

# Evaluate the Linear Regression model
print("Linear Regression Mean Squared Error:", mean_squared_error(y_test, y_pred_linear))
print("Linear Regression R^2 Score:", r2_score(y_test, y_pred_linear))

Linear Regression Mean Squared Error: 57.112469807680824
Linear Regression R^2 Score: 0.6882549296614693


Random Forest Regression

In [19]:


# Initialize and train the Random Forest Regressor
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_rf = rf_model.predict(X_test)

# Evaluate the Random Forest Regressor
print("Random Forest Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred_rf))
print("Random Forest Regressor R^2 Score:", r2_score(y_test, y_pred_rf))

Random Forest Regressor Mean Squared Error: 26.168811363636372
Random Forest Regressor R^2 Score: 0.8571590763505131


Gradient Boosting Model

In [21]:
#define the gradient boosting model with different parameters
gbm_model = GradientBoostingRegressor(n_estimators=300, learning_rate=0.1, max_depth=3, random_state=42)
gbm_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_gbm = gbm_model.predict(X_test)

# Evaluate the Gradient Boosting Regressor
print("Gradient Boosting Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred_gbm))
print("Gradient Boosting Regressor R^2 Score:", r2_score(y_test, y_pred_gbm))


Gradient Boosting Regressor Mean Squared Error: 23.454100517656478
Gradient Boosting Regressor R^2 Score: 0.8719771664537529


Ridge Regression

In [25]:

ridge = Ridge(alpha=0.1)
ridge.fit(X_train, y_train)
y_pred_ridge = ridge.predict(X_test)

print("Gradient Boosting Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred_ridge))
print("Gradient Boosting Regressor R^2 Score:", r2_score(y_test, y_pred_ridge))

Gradient Boosting Regressor Mean Squared Error: 57.14544675473055
Gradient Boosting Regressor R^2 Score: 0.6880749269280526


CONCLUSION

Linear Regression - 0.688

Random Forest Regressor - 0.857

Gradient Boosting Regressor - 0.871

Ridge Regression - 0.688
